source/Cosmos.HAL2/BlockDevice/AHCI.cs

```diff
@@ -0,0 +1,380 @@
+            SupportsSNotificationRegister         = (mGeneric.Capabilities >> 29 & 1) == 1;
+            SupportsNativeCommandQueuing          = (mGeneric.Capabilities >> 30 & 1) == 1;
+            Supports64bitAddressing               = (mGeneric.Capabilities >> 31 & 1) == 1;
+        }
+
+        private void GetPorts()
+        {
+            // Search for disks
+            var xImplementedPort = mGeneric.ImplementedPorts;
+            var xPort = 0;
+            for(; xPort < 32; xPort++)
+            {
+                if ((xImplementedPort & 1) != 0)
+                {
+                    PortRegisters xPortReg = new PortRegisters((uint)mABAR + 0x100, (uint)xPort);
+                    PortType PortType = CheckPortType(xPortReg);
+                    xPortReg.mPortType = PortType;
+                    var xPortString = "0:" + ((xPort.ToString().Length <= 1) ? xPort.ToString().PadLeft(1, '0') : xPort.ToString());
+                    if (PortType == PortType.SATA) // If Port type was SATA.
+                    {
+                        mAHCIDebugger.Send("Initializing Port " + xPortString + " with type SATA");
+                        PortRebase(xPortReg, (uint)xPort);
+                        var xSATAPort = new SATA(xPortReg);
+                        mPorts.Add(xSATAPort);
+                    }
+                    else if (PortType == PortType.SATAPI) // If Port type was SATAPI.
+                    {
+                        mAHCIDebugger.Send("Initializing Port " + xPortString + " with type Serial ATAPI");
+                        //PortRebase(xPortReg, (uint)xPort);
+                        //var xSATAPIPort = new SATAPI(xPortReg);
+                        //mPorts.Add(xSATAPIPort);
+                    }
+                    else if (PortType == PortType.SEMB) // If Port type was SEMB.
+                    {
+                        mAHCIDebugger.Send("SEMB Drive at port " + xPortString + " found, which is not supported yet!");
+                    }
+                    else if (PortType == PortType.PM) // If Port type was Port Mulitplier.
+                    {
+                        mAHCIDebugger.Send("Port Multiplier Drive at port " + xPortString + " found, which is not supported yet!");
+                    }
+                    else if (PortType != PortType.Nothing)
+                        mAHCIDebugger.Send("Unknown drive found with signature: 0x" + xPortReg.SIG);// If Implemented Port value was not zero and non of the above.
+                }
+                xImplementedPort >>= 1;
+            }
+        }
+
+        private PortType CheckPortType(PortRegisters aPort)
+        {
+            var xIPM = (InterfacePowerManagementStatus)((aPort.SSTS >> 8) & 0x0F);
+            var xSPD = (CurrentInterfaceSpeedStatus)((aPort.SSTS >> 4) & 0x0F);
+            var xDET = (DeviceDetectionStatus)(aPort.SSTS & 0x0F);
+            var xSignature = aPort.SIG;
+
+            // Check if not reading wrong data!
```

Comment:
Usually when something "impossible" happens in C# better throw an exception instead to return a "default" value.